eeprom.c

```diff
@@ -24,7 +24,7 @@ static int is_eeprom(const char *fpath, const struct stat *sb,
 	char *fpath_s = strdup(fpath);
 	if (typeflag == FTW_F && !strcmp(basename(fpath_s), "eeprom") \
 			&& sb->st_size == FAB_SIZE_FRU_EEPROM) {
-		free(eeprom_path);
```

Comment:
typically, `free()` handles NULL; at least for glibc;
is there a case where this isn't happening ?
i admit that there are some libc implementations that may not handle that ;

but, if you want, you could just do a NULL check;
otherwise this may risk doing leaks

Reply:
Yes, you are right, `free()` should handle it correctly. I'm not sure if it is applicable to osc, but in multithreaded applications, that free may be an obvious problem leaving some thread pointing to memory which was freed already... Either way, typically in this API's such as `find_eeprom()` it's the caller responsibility to free the allocated memory. But, if you prefer and if the multithreaded case is not applicable to osc i can just do the null check...

Reply:
ok, but if the discussion is about multithreaded stuff, then the fix is to add some sort of locking there, since it's a shared resource

or to rework it, so that it's not shared;

Reply:
> or to rework it, so that it's not shared;
That would be the best option in my opinion. But since we are using `nftw()` might be not possible. I will check if we can rework it...

Reply:
don't invest too much energy into this;
locking may be good enough as well;

Reply:
Ok. The naming for the PR is not that correct. Indeed, there is no problem with `free(NULL)`, but assuming that the caller is freeing the memory, calling free again on already free'd memory leads to undefined behavior. For removing the shared resource, the only alternative i could come with, is doing something like [here](https://github.com/analogdevicesinc/iio-oscilloscope/blob/0573b1e6883ac21f8d7049fc3ae33312bbf13ae7/plugins/fmcomms1.c#L1403). Honestly, Im not a big fan of `popen()` and `system()` but it saves us from a shared resource. The other way would be too time consuming (using directly the API's in dirent.h). I will update the PR with the reworked function...

Reply:
i don't see how double-free could happen, unless there is some weird multi-threading bug ;
`eeprom_path` is re-assigned right after being free'd;

`popen() & system()` seem a bit overkill for this ; but i wouldn't mind them much here;

using `dirent.h` shouldn't be too bad either (time consuming);
i mean, it takes 30-50 lines to write a function that could be generic;

Reply:
> i don't see how double-free could happen

I think it would happen in the second time we make the call to `find_eeprom()`. We return the address of the allocated memory with `strdup()` to the caller which typically will call `free()` or `g_free()` ([here)](https://github.com/analogdevicesinc/iio-oscilloscope/blob/0573b1e6883ac21f8d7049fc3ae33312bbf13ae7/plugins/fmcomms2.c#L873),leaving our shared resource in `eeprom.c` pointing to an already freed region...

> but i wouldn't mind them much here;

Then if it is acceptable I will go for it since I already have a proposal...

>using dirent.h shouldn't be too bad either;

I'm just worried with this because we would need some recursion here to have the recursive search for a given directory which can be not that straight... But if you prefer this I can invest a bit more on it...

Reply:
>  I think it would happen in the second time we make the call to find_eeprom(). We return the address of the allocated memory with strdup() to the caller which typically will call free() or g_free() (here),leaving our shared resource in eeprom.c pointing to an already freed region...

oh, i just saw that;
so, just remove the static pointer `static char *eeprom_path = NULL;` and make sure that it is free'd wherever `find_eeprom()` is called

Reply:
Yes, all the callers are doing their job :)